### PR TITLE
Increasing Test Coverage in Ledger Rules

### PR DESCRIPTION
**Pull request by yHSJ**
Title describes all

## Summary by CodeRabbit

- **New Features**
  - Added a new error for collateral return overflow in transaction validation.
  - Extended support for transaction outputs to include an optional value field.
  - Enhanced tracing for transaction validation steps, providing more detailed runtime observability.

- **Bug Fixes**
  - Corrected error message formatting for input validation errors.

- **Tests**
  - Added comprehensive test cases for transaction validation, including collateral, witness, input, withdrawal, and voting procedure scenarios.
  - Introduced extensive new test data covering various transaction edge cases and validation outcomes.

- **Style**
  - Minor formatting improvements for code readability.

### CODE REVIEW

File: crates/amaru-ledger/src/context/assert.rs [hunk 1]
```diff
@@ -252,6 +269,14 @@ impl WitnessSlice for AssertValidationContext {
         }
     }
 
+    #[instrument(
+        level = Level::TRACE,
+        fields(
+            bootstrap_witness.hash = %root,
+        )
+        skip_all,
+        name = "require_bootstrap_witness"
```
Comment (KtorZ) [inline crates/amaru-ledger/src/context/assert.rs:278]: Note: I don't think you need to specify the name when it's identical to the function being instrumented.

File: crates/amaru-ledger/src/rules/transaction/fees.rs [hunk 2]
```diff
@@ -19,6 +19,11 @@ use amaru_kernel::{HasLovelace, Lovelace, MintedTransactionOutput, TransactionIn
 pub enum InvalidFees {
     #[error("unknown collateral input at position {position}")]
     UnknownCollateralInput { position: usize },
+    #[error("Collateral return value is greater than total collateral input (input: {total_collateral_input}, return: {total_collateral_return})")]
```
Comment (KtorZ) [inline crates/amaru-ledger/src/rules/transaction/fees.rs]:
```suggestion
    #[error("collateral return value ({total_collateral_return}) is greater than total collateral input ({total_collateral_input})")]
```

1. Start with a lowercase, so that we can more easily chain error messages: `2nd transaction failed: invalid fees: collateral return...`. Although, we should really think about producing _structured_ errors in the long run. 

2. Embed information where it's relevant; avoid repetition and makes the error message easier to read.

File: crates/amaru-ledger/src/rules/transaction/inputs.rs [hunk 3]
```diff
@@ -18,14 +18,14 @@ use thiserror::Error;
 
 #[derive(Debug, Error)]
 pub enum InvalidInputs {
-    #[error("Unknown input: {}#{}", .0.index, .0.transaction_id)]
+    #[error("Unknown input: {}#{}", .0.transaction_id, .0.index)]
```
Comment (KtorZ) [inline crates/amaru-ledger/src/rules/transaction/inputs.rs:21]: :'( 

I know this is how cardano-cli typically shows output reference, but it's a terrible convention IMO. Having the index far at the back makes it easy to miss. But, I won't fight for it (now).

File: crates/amaru-ledger/src/rules/transaction/inputs.rs [hunk 4]
```diff
@@ -117,30 +120,67 @@ mod tests {
...
-    fn valid_reference_input_transactions(
+    #[test_case(fixture!("7a098c13f3fb0119bc1ea6a418af3b9b8fef18bb65147872bf5037d28dda7b7b", "non-disjoint-reference-inputs") =>
+        matches Err(InvalidInputs::NonDisjointRefInputs { intersection })
+            if  intersection.len() == 1
+                && InvalidInputs::NonDisjointRefInputs { intersection: intersection.clone() }.to_string() == "inputs included in both reference inputs and spent inputs: intersection [47a890217e4577ec3e6d5db161a4aa524a5cce3302e389ccb22b5662146f52ab#0]";
+        "Non-Disjoint Reference Inputs")]
+    #[test_case(fixture!("7a098c13f3fb0119bc1ea6a418af3b9b8fef18bb65147872bf5037d28dda7b7b", "invalid-address-header") =>
+        matches Err(InvalidInputs::UncategorizedError(e))
+            if e == "Invalid output address. (error InvalidHeader(160)) output: PostAlonzo(PseudoPostAlonzoTransactionOutput { address: Bytes(ByteVec([160, 83, 9, 250, 120, 104, 86, 193, 38, 45, 9, 91, 137, 173, 246, 79, 232, 165, 37, 90, 209, 145, 66, 201, 197, 55, 53, 158, 65])), value: Coin(0), datum_option: None, script_ref: None })";
```
Comment (KtorZ) [inline crates/amaru-ledger/src/rules/transaction/inputs.rs]:
```suggestion
            if e.contains("InvalidHeader")
```

File: crates/amaru-ledger/src/rules/transaction/inputs.rs [hunk 5]
```diff
@@ -117,30 +120,67 @@ mod tests {
...
+            "invalid address header")]
+    #[test_case(fixture!("7a098c13f3fb0119bc1ea6a418af3b9b8fef18bb65147872bf5037d28dda7b7b", "unknown-input") =>
+        matches Err(InvalidInputs::UnknownInput(input))
+        if  hex::encode(input.transaction_id) == "47a890217e4577ec3e6d5db161a4aa524a5cce3302e389ccb22b5662146f52ab" && input.index == 2;
+        "unknown input")]
+    #[test_case(fixture!("7a098c13f3fb0119bc1ea6a418af3b9b8fef18bb65147872bf5037d28dda7b7b", "invalid-byron-address") =>
+        matches Err(InvalidInputs::UncategorizedError(e))
+        if e == "Invalid byron address payload. (error InvalidByronCbor(Error { err: TypeMismatch(U8), pos: Some(31), msg: \"expected map\" })) address: ByronAddress { payload: TagWrap(ByteVec([130, 88, 28, 133, 24, 18, 154, 60, 13, 248, 227, 60, 64, 224, 75, 141, 38, 173, 59, 4, 34, 208, 250, 156, 169, 37, 88, 6, 163, 243, 139, 0])), crc: 3884043611 }";
```
Comment (KtorZ) [inline crates/amaru-ledger/src/rules/transaction/inputs.rs]:
```suggestion
        if e.contains("InvalidByronCbor");
```

**Review by KtorZ — APPROVED** [review]
I didn't inspect all traces _in details_ but overall looks good.

File: crates/amaru-ledger/src/context/assert.rs (hunk 1 above)
Comment (yHSJ) [reply crates/amaru-ledger/src/context/assert.rs:278]: You may be right, but I like specifying it -- in case something changes or whatever
